support for Gisaid upload

mayabrandi opened the pull request:
## Description 
*This PR adds support for Gisaid upload.*

**How to prepare for test**:
- [x] ssh to hasta
- [x] then:

```
up
bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh gisaid
bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-servers-stage.sh gisaid
us
python /home/proj/stage/servers/resources/gisaid_uploader.py CoV authenticate --cid TEST-EA76875B00C3
cg -c /home/proj/stage/servers/config/hasta.scilifelab.se/cg-stage.yaml upload gisaid frankhusky
```

**Expected test outcome**:
- [x] check thet the gisaid upload works as expected

**Review:**
- [x] code approved by @Mropat 
- [x] tests executed by @mayabrandi 
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!


### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch
- [ ] Inform to ...

moonso commented on cg/cli/upload/gisaid.py:
```diff
@@ -0,0 +1,41 @@
+
+    meta_api = MetaAPI(context.obj)
+    status_db = meta_api.status_db
+    click.echo(click.style("----------------- GISAID -------------------"))
```
```suggestion
   LOg.info("----------------- GISAID -------------------")
```

mayabrandi replied:
👍

moonso commented on cg/cli/upload/gisaid.py:
```diff
@@ -0,0 +1,41 @@
+    """Upload mutant analysis data to GISAID."""
+
+    meta_api = MetaAPI(context.obj)
+    status_db = meta_api.status_db
```
Type hint

mayabrandi replied:
👍

moonso commented on cg/cli/upload/gisaid.py:
```diff
@@ -0,0 +1,41 @@
+    status_db = meta_api.status_db
+    click.echo(click.style("----------------- GISAID -------------------"))
+
+    case_object = status_db.family(family_id)
```
type hint

mayabrandi replied:
👍

moonso commented on cg/cli/upload/gisaid.py:
```diff
@@ -0,0 +1,41 @@
+
+    case_object = status_db.family(family_id)
+    ticket_number = case_object.name
+    samples: List[models.Sample] = status_db.Sample.query.filter_by(
```
Please use the functions that exist in the API (the status_db api) or add the functionality there if it does not exist. Code like this will be hard to hunt down and change when we change the API

mayabrandi replied:
That was the plan 👍

moonso commented on cg/cli/upload/gisaid.py:
```diff
@@ -0,0 +1,41 @@
+    samples: List[models.Sample] = status_db.Sample.query.filter_by(
+        ticket_number=ticket_number
+    ).all()
+    upload_gisaid_api = UploadGisaidAPI(
```
type hint

moonso commented on cg/cli/upload/gisaid.py:
```diff
@@ -0,0 +1,41 @@
+        family_id=family_id,
+    )
+
+    files: dict = upload_gisaid_api.files()
```
pedantic class?

mayabrandi commented on cg/cli/upload/gisaid.py:
```diff
@@ -0,0 +1,41 @@
+    click.echo(click.style("----------------- GISAID -------------------"))
+
+    case_object = status_db.family(family_id)
+    ticket_number = case_object.name
```
👍

moonso commented on cg/apps/gisaid/gisaid.py:
```diff
@@ -0,0 +1,125 @@
+from typing import List
+
+from cg.store import models
+from ...meta.workflow.fastq import FastqHandler
```
change to absolute imports
```suggestion
from cg.meta.workflow.fastq import FastqHandler
```

mayabrandi commented on cg/meta/workflow/fastq.py:
```diff
@@ -94,6 +94,14 @@ def get_concatenated_name(linked_fastq_name: str) -> str:
 
+    @staticmethod
+    def get_header(fastq_path: Path) -> str:
+        """Get header from fasta file"""
```
```suggestion
        """Get header from fasta file. Assuming un zipped file"""
```

mayabrandi commented on cg/meta/workflow/fastq.py:
```diff
@@ -94,6 +94,14 @@ def get_concatenated_name(linked_fastq_name: str) -> str:
+    @staticmethod
+    def get_header(fastq_path: Path) -> str:
+        """Get header from fasta file"""
+        # HAndel if ziped file??
```
```suggestion
```

mayabrandi commented on cg/store/api/findbusinessdata.py:
```diff
@@ -132,6 +132,18 @@ def family_samples(self, family_id: str) -> List[models.FamilySample]:
         )
 
+    def get_samples_by_family_id(self, family_id: str) -> List[models.Sample]:
+        """Get samples on a given case-id"""
```
```suggestion
        """Get samples on a given family_id"""
```

Mropat commented on cg/apps/gisaid/constants.py:
```diff
@@ -0,0 +1,16 @@
+AUTHORS = [
```
We still need these?

mayabrandi replied:
yep

Mropat commented on cg/apps/gisaid/gisaid.py:
```diff
@@ -0,0 +1,126 @@
+    def get_sample_row(self, sample: models.Sample, family_id: str, fasta_header: str) -> List[str]:
+        """Build row for a sample in the batch upload csv."""
+
+        orig_lab: str = "Stockholm"  # sample.originating_lab
```
can we just use those values as defaults in the model? instead of passing

mayabrandi replied:
No. These values are sample dependent and deed to be stored in the database. We are waiting for the clinichasn to find a solution for how to pass the info via the order. Its not solved yet. Therefor hardcoded now.

Mropat commented on cg/apps/gisaid/gisaid.py:
```diff
@@ -0,0 +1,126 @@
+            covv_collection_date=collection_date,
+            lab=orig_lab,
+        )
+        return [gisaid_sample.dict().get(header) for header in self.get_headers()]
```
if get headers are same sorting order as gisaid_sample.dict, does it mean you only want a list of dict values?

mayabrandi replied:
Yes. I want a list of dict values in a specific order

moonso replied:
Agreed on going back to use HEADER constant

Mropat commented on cg/apps/gisaid/gisaid.py:
```diff
@@ -0,0 +1,126 @@
+        self.fastq_handler.concatenate(files=fasta_files, concat_file=file_name)
+        return file_name
+
+    def files(self, family_id: str) -> UpploadFiles:
```
is this method needed? could move the get_file methods directly to upload! Or i would do so, you decide! its a preference

mayabrandi replied:
👍

moonso commented on cg/apps/gisaid/gisaid.py:
```diff
@@ -0,0 +1,126 @@
+
+from cg.store import models
+from cg.meta.workflow.fastq import FastqHandler
+from ...utils import Process
```
I have the solution to this, can show you later if you are interested

mayabrandi replied:
Yes please

moonso commented on cg/apps/gisaid/gisaid.py:
```diff
@@ -0,0 +1,126 @@
+    def get_headers(self) -> list:
+        """Get GisaidSample keys as list."""
+
+        return list(GisaidSample.schema().get("properties").keys())
```
Cool I've never seen this!

moonso replied:
Would this be the same as doing 
```
list(GisaidSample.dict().keys())
```
? Do you know @mayabrandi or @Mropat ?

moonso replied:
Or what would be the difference?

Mropat replied:
@moonso I think it might be the same! but you need an instance of Gisaidsample() with valid values. This gets you the expected properties without there being an instance of class

mayabrandi replied:
I guess the GisaidSample.dict() will lose the order of the fields. 
GisaidSample.schema() will keep the order of the fields.

moonso commented on cg/apps/gisaid/gisaid.py:
```diff
@@ -0,0 +1,126 @@
+        """Build row for a sample in the batch upload csv."""
+
+        orig_lab: str = "Stockholm"  # sample.originating_lab
+        collection_date: str = "201122"  # sample.collection_date str or date, we dont know yet
```
Let's try to use ISO formated dates as much as possible. That will make it easier for every library. (YYYY-MM-DD)

mayabrandi replied:
is type str then or datetime.date.isorformat typ?

Mropat commented on cg/apps/gisaid/gisaid.py:
```diff
@@ -0,0 +1,126 @@
+
+from cg.meta.meta import MetaAPI
+
+LOG = logging.getLogger(__name__)
```
sort logging to be after imports?

moonso commented on cg/apps/gisaid/gisaid.py:
```diff
@@ -0,0 +1,124 @@
+from cg.meta.meta import MetaAPI
+import csv
+
+from ...exc import HousekeeperVersionMissing
```
use absolute import

moonso commented on cg/apps/gisaid/gisaid.py:
```diff
@@ -0,0 +1,124 @@
+        hk_version: hk_models.Version = self.housekeeper_api.last_version(bundle=family_id)
+        if not hk_version:
+            LOG.info("Family ID: %s not found in hose keeper", family_id)
+            raise HousekeeperVersionMissing()
```
```suggestion
            raise HousekeeperVersionMissingError()
```
Super!

moonso commented on cg/exc.py:
```diff
@@ -131,6 +131,12 @@ class MandatoryFilesMissing(CgError):
     """
 
 
+class HousekeeperVersionMissing(CgError):
```
```suggestion
class HousekeeperVersionMissingError(CgError):
```

Mropat commented on cg/meta/upload/gisaid/gisaid.py:
```diff
@@ -0,0 +1,136 @@
+            fasta_file: str = self.housekeeper_api.files(
+                version=hk_version.id, tags=["consensus", sample.cg_lims_id]
+            ).first()
+            fasta_file = "/Users/maya.brandi/opt/cg/f1.fasta"
```
this will give you just your local file

mayabrandi replied:
:)

Mropat commented on cg/meta/upload/gisaid/gisaid.py:
```diff
@@ -0,0 +1,136 @@
+        self.gisaid_binary: str = config.gisaid.binary_path
+        self.process = Process(binary=self.gisaid_binary)
+
+    def get_fasta_sequence(self, fastq_path: str) -> str:
```
Suggest maybe call this get_fasta_header?

Mropat commented on cg/meta/upload/gisaid/gisaid.py:
```diff
@@ -0,0 +1,136 @@
+
+        fasta_objects = []
+        for sample in gsaid_samples:
+            hk_version: hk_models.Version = self.housekeeper_api.last_version(
```
you can fetch housekeeper version once, since its the same bundle name for all files

Mropat replied:
I mean, if you fetch it before the loop

Mropat commented on cg/store/api/findbusinessdata.py:
```diff
@@ -132,6 +132,18 @@ def family_samples(self, family_id: str) -> List[models.FamilySample]:
+        """Get sequenced samples by family_id"""
+
+        samples: List[models.Sample] = self.get_samples_by_family_id(family_id)
+        return [sample for sample in samples if sample.sequenced_at]
```
The logic with sequenced_at update recently changed. 
can you use `if sample_obj.reads > sample_obj.application_version.application.expected_reads`

Mropat replied:
or add a property on Sample table to check if this threshold is passed

mayabrandi replied:
okej! 👍

Mropat commented on gisaid_uploader.authtoken:
```diff
@@ -0,0 +1 @@
+{"CoV": "TEST-EA76875B00C3/OOLIONENRPATOPLSSNQLGXJSYATQZMMCCTRBPKHKEOVNHKUCOJWGKNCAFNMSFKZLYDABBHXCQNHDTLERDZOKZUHYXCGJSLZUVHJNEIDHIGOSWZFBQMNURKSMJWSENTBL"}
```
I hope this is public info? I suggest put the authtoken in servers config

Mropat commented on cg/cli/upload/gisaid.py:
```diff
@@ -0,0 +1,37 @@
+        raise click.Abort
+
+    gisaid_api = GisaidAPI(config=context)
+    gsaid_samples: List[GisaidSample] = gisaid_api.get_gsaid_samples(family_id=family_id)
+    files: UpploadFiles = UpploadFiles(
+        csv_file=gisaid_api.build_gisaid_csv(gsaid_samples=gsaid_samples),
+        fasta_file=gisaid_api.build_gisaid_fasta(gsaid_samples=gsaid_samples),
+    )
+    if files:
```
I would move all logic into GisaidAPI
so that cli only has gisaid_api.upload(case_id)
and rest of logic resolved in the class

Mropat commented on gisad.csv:
```diff
@@ -0,0 +1,10 @@
+submitter,fn,covv_virus_name,covv_type,covv_passage,covv_collection_date,covv_location,covv_host,covv_gender,covv_patient_age,covv_patient_status,covv_seq_technology,covv_orig_lab,covv_orig_lab_addr,covv_subm_sample_id,covv_subm_lab,covv_subm_lab_addr,covv_authors,covv_specimen,covv_outbreak,covv_add_host_info,covv_add_location,covv_provider_sample_id,covv_last_vaccinated,covv_treatment,covv_asse ...
```
dont forget to remove test outputs! And make sure the files are stored someplace in mutant result directory!

mayabrandi commented:
Testing on hasta:


````

[0|0|524] 2d [maya.brandi@hasta:~] [P_main] $ bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh gisaid
on the host hasta.scilifelab.se, continue
/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Collecting git+https://github.com/Clinical-Genomics/cg@gisaid
  Cloning https://github.com/Clinical-Genomics/cg (to revision gisaid) to ./tmp/pip-req-build-k474n6_3
  Running command git clone -q https://github.com/Clinical-Genomics/cg /home/maya.brandi/tmp/pip-req-build-k474n6_3
  Running command git checkout -b gisaid --track origin/gisaid
  Switched to a new branch 'gisaid'
  Branch gisaid set up to track remote branch gisaid from origin.
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Building wheels for collected packages: cg
  Building wheel for cg (PEP 517) ... done
  Created wheel for cg: filename=cg-20.22.0-py3-none-any.whl size=438149 sha256=86c924a39548be01b2f22d34f091e95e217e0337fc35a36a24202a8ba3b6a56a
  Stored in directory: /home/maya.brandi/tmp/pip-ephem-wheel-cache-r7hkwjwg/wheels/48/80/2c/ebbd2f09c189342739a9cff408f2f02ba0346175ff0906c219
Successfully built cg
Installing collected packages: cg
  Attempting uninstall: cg
    Found existing installation: cg 20.22.0
    Uninstalling cg-20.22.0:
      Successfully uninstalled cg-20.22.0
Successfully installed cg-20.22.0
Collecting git+https://github.com/Clinical-Genomics/cg@gisaid
  Cloning https://github.com/Clinical-Genomics/cg (to revision gisaid) to ./tmp/pip-req-build-255ws7en
  Running command git clone -q https://github.com/Clinical-Genomics/cg /home/maya.brandi/tmp/pip-req-build-255ws7en
  Running command git checkout -b gisaid --track origin/gisaid
  Switched to a new branch 'gisaid'
  Branch gisaid set up to track remote branch gisaid from origin.
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: Flask>=1.1.1 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (1.1.2)
Requirement already satisfied: Click<7.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (6.7)
Requirement already satisfied: Flask-Admin in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (1.5.3)
Requirement already satisfied: openpyxl in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (2.6.0)
Requirement already satisfied: Flask-Alchy in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (0.5.0)
Requirement already satisfied: lxml in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (4.3.1)
Requirement already satisfied: pyschemes in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (1.0.0)
Requirement already satisfied: genologics in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (0.4.6)
Requirement already satisfied: tabulate in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (0.8.3)
Requirement already satisfied: marshmallow>3 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (3.6.1)
Requirement already satisfied: typing-extensions in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (3.7.4.3)
Requirement already satisfied: pydantic in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (1.8.1)
Requirement already satisfied: colorclass in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (2.2.0)
Requirement already satisfied: requests[security] in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (2.21.0)
Requirement already satisfied: Flask-Dance in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (1.3.0)
Requirement already satisfied: Alchy in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (2.2.2)
Requirement already satisfied: requests-cache in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (0.4.13)
Requirement already satisfied: setuptools>=39.2.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (40.6.2)
Requirement already satisfied: Flask-SQLAlchemy==2.1 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (2.1)
Requirement already satisfied: cgstats>=1.1.2 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (1.7.0)
Requirement already satisfied: coloredlogs in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (14.0)
Requirement already satisfied: Flask-CORS in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (3.0.7)
Requirement already satisfied: python-dateutil in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (2.8.0)
Requirement already satisfied: petname in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (2.2)
Requirement already satisfied: google-auth in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (1.6.2)
Requirement already satisfied: blinker in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (1.4)
Requirement already satisfied: gunicorn in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (20.0.4)
Requirement already satisfied: email-validator in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (1.1.2)
Requirement already satisfied: SQLAlchemy<1.4 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (1.3.23)
Requirement already satisfied: werkzeug<1.0.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (0.16.1)
Requirement already satisfied: ruamel.yaml in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (0.15.88)
Requirement already satisfied: cgmodels>=0.5 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (0.5.0)
Requirement already satisfied: housekeeper>=2.5 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cg==20.22.0) (3.2.0)
Requirement already satisfied: Flask-Bootstrap in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (3.3.7.1)
Requirement already satisfied: docutils in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (0.16)
Requirement already satisfied: PyYAML>=5.3.1 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (5.3.1)
Requirement already satisfied: yamllint in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (1.25.0)
Requirement already satisfied: pylint in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (2.6.0)
Requirement already satisfied: alembic in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (1.0.7)
Requirement already satisfied: wheel in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (0.32.3)
Requirement already satisfied: pycodestyle in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (2.6.0)
Requirement already satisfied: path.py in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (12.5.0)
Requirement already satisfied: six>=1.12 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (1.15.0)
Requirement already satisfied: pytest in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (4.2.1)
Requirement already satisfied: bs4 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (0.0.1)
Requirement already satisfied: pymysql in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (0.9.3)
Requirement already satisfied: html-linter in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (0.4.0)
Requirement already satisfied: git-lint in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (0.1.2)
Requirement already satisfied: bumpversion in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cgstats>=1.1.2->cg==20.22.0) (0.5.3)
Requirement already satisfied: itsdangerous>=0.24 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Flask>=1.1.1->cg==20.22.0) (1.1.0)
Requirement already satisfied: Jinja2>=2.10.1 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Flask>=1.1.1->cg==20.22.0) (2.11.3)
Requirement already satisfied: rich in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from housekeeper>=2.5->cg==20.22.0) (6.0.0)
Requirement already satisfied: MarkupSafe>=0.23 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Jinja2>=2.10.1->Flask>=1.1.1->cg==20.22.0) (1.1.1)
Requirement already satisfied: pydash>=3.4.7 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Alchy->cg==20.22.0) (4.9.2)
Requirement already satisfied: Mako in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from alembic->cgstats>=1.1.2->cg==20.22.0) (1.0.7)
Requirement already satisfied: python-editor>=0.3 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from alembic->cgstats>=1.1.2->cg==20.22.0) (1.0.4)
Requirement already satisfied: beautifulsoup4 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from bs4->cgstats>=1.1.2->cg==20.22.0) (4.7.1)
Requirement already satisfied: soupsieve>=1.2 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from beautifulsoup4->bs4->cgstats>=1.1.2->cg==20.22.0) (2.2)
Requirement already satisfied: humanfriendly>=7.1 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from coloredlogs->cg==20.22.0) (8.2)
Requirement already satisfied: dnspython>=1.15.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from email-validator->cg==20.22.0) (2.1.0)
Requirement already satisfied: idna>=2.0.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from email-validator->cg==20.22.0) (2.8)
Requirement already satisfied: wtforms in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Flask-Admin->cg==20.22.0) (2.2.1)
Requirement already satisfied: dominate in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Flask-Bootstrap->cgstats>=1.1.2->cg==20.22.0) (2.3.5)
Requirement already satisfied: visitor in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Flask-Bootstrap->cgstats>=1.1.2->cg==20.22.0) (0.1.3)
Requirement already satisfied: oauthlib in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Flask-Dance->cg==20.22.0) (2.1.0)
Requirement already satisfied: urlobject in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Flask-Dance->cg==20.22.0) (2.4.3)
Requirement already satisfied: requests-oauthlib>=1.0.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Flask-Dance->cg==20.22.0) (1.1.0)
Requirement already satisfied: lazy in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from Flask-Dance->cg==20.22.0) (1.4)
Requirement already satisfied: certifi>=2017.4.17 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from requests[security]->cg==20.22.0) (2018.10.15)
Requirement already satisfied: urllib3<1.25,>=1.21.1 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from requests[security]->cg==20.22.0) (1.24.3)
Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from requests[security]->cg==20.22.0) (3.0.4)
Requirement already satisfied: docopt in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from git-lint->cgstats>=1.1.2->cg==20.22.0) (0.6.2)
Requirement already satisfied: pathlib2 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from git-lint->cgstats>=1.1.2->cg==20.22.0) (2.3.5)
Requirement already satisfied: termcolor in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from git-lint->cgstats>=1.1.2->cg==20.22.0) (1.1.0)
Requirement already satisfied: template-remover in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from git-lint->cgstats>=1.1.2->cg==20.22.0) (0.1.9)
Requirement already satisfied: cachetools>=2.0.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from google-auth->cg==20.22.0) (3.1.0)
Requirement already satisfied: rsa>=3.1.4 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from google-auth->cg==20.22.0) (4.0)
Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from google-auth->cg==20.22.0) (0.2.4)
Requirement already satisfied: pyasn1<0.5.0,>=0.4.1 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from pyasn1-modules>=0.2.1->google-auth->cg==20.22.0) (0.4.5)
Requirement already satisfied: jdcal in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from openpyxl->cg==20.22.0) (1.4)
Requirement already satisfied: et-xmlfile in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from openpyxl->cg==20.22.0) (1.0.1)
Requirement already satisfied: path in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from path.py->cgstats>=1.1.2->cg==20.22.0) (15.0.1)
Requirement already satisfied: isort<6,>=4.2.5 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from pylint->cgstats>=1.1.2->cg==20.22.0) (5.6.4)
Requirement already satisfied: mccabe<0.7,>=0.6 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from pylint->cgstats>=1.1.2->cg==20.22.0) (0.6.1)
Requirement already satisfied: astroid<=2.5,>=2.4.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from pylint->cgstats>=1.1.2->cg==20.22.0) (2.4.2)
Requirement already satisfied: toml>=0.7.1 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from pylint->cgstats>=1.1.2->cg==20.22.0) (0.10.0)
Requirement already satisfied: wrapt~=1.11 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from astroid<=2.5,>=2.4.0->pylint->cgstats>=1.1.2->cg==20.22.0) (1.12.1)
Requirement already satisfied: lazy-object-proxy==1.4.* in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from astroid<=2.5,>=2.4.0->pylint->cgstats>=1.1.2->cg==20.22.0) (1.4.3)
Requirement already satisfied: typed-ast<1.5,>=1.4.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from astroid<=2.5,>=2.4.0->pylint->cgstats>=1.1.2->cg==20.22.0) (1.4.1)
Requirement already satisfied: atomicwrites>=1.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from pytest->cgstats>=1.1.2->cg==20.22.0) (1.3.0)
Requirement already satisfied: more-itertools>=4.0.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from pytest->cgstats>=1.1.2->cg==20.22.0) (6.0.0)
Requirement already satisfied: attrs>=17.4.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from pytest->cgstats>=1.1.2->cg==20.22.0) (18.2.0)
Requirement already satisfied: pluggy>=0.7 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from pytest->cgstats>=1.1.2->cg==20.22.0) (0.8.1)
Requirement already satisfied: py>=1.5.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from pytest->cgstats>=1.1.2->cg==20.22.0) (1.7.0)
Requirement already satisfied: pyOpenSSL>=0.14 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from requests[security]->cg==20.22.0) (19.0.0)
Requirement already satisfied: cryptography>=1.3.4 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from requests[security]->cg==20.22.0) (2.5)
Requirement already satisfied: asn1crypto>=0.21.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cryptography>=1.3.4->requests[security]->cg==20.22.0) (0.24.0)
Requirement already satisfied: cffi!=1.11.3,>=1.8 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cryptography>=1.3.4->requests[security]->cg==20.22.0) (1.14.5)
Requirement already satisfied: pycparser in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from cffi!=1.11.3,>=1.8->cryptography>=1.3.4->requests[security]->cg==20.22.0) (2.20)
Requirement already satisfied: pygments<3.0.0,>=2.6.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from rich->housekeeper>=2.5->cg==20.22.0) (2.6.1)
Requirement already satisfied: colorama<0.5.0,>=0.4.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from rich->housekeeper>=2.5->cg==20.22.0) (0.4.3)
Requirement already satisfied: commonmark<0.10.0,>=0.9.0 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from rich->housekeeper>=2.5->cg==20.22.0) (0.9.1)
Requirement already satisfied: pathspec>=0.5.3 in /home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages (from yamllint->cgstats>=1.1.2->cg==20.22.0) (0.7.0)
/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
remote: Enumerating objects: 82, done.
remote: Counting objects: 100% (82/82), done.
remote: Compressing objects: 100% (30/30), done.
remote: Total 82 (delta 54), reused 79 (delta 52), pack-reused 0
Unpacking objects: 100% (82/82), done.
From github.com:Clinical-Genomics/cg
   6e25436..67a584b  feat-trended-cases -> origin/feat-trended-cases
   a0e377d..2d35ad2  gisaid     -> origin/gisaid
   83dbd88..f60ba89  master     -> origin/master
 * [new tag]         v20.22.0   -> v20.22.0
Switched to branch 'master'
Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
Updating 83dbd88..f60ba89
Fast-forward
 .bumpversion.cfg                       |  2 +-
 CHANGELOG.md                           |  5 ++++-
 cg/cli/get.py                          | 34 +++++++++++++++++++++++++++++++---
 setup.py                               |  2 +-
 tests/cli/get/test_cli_get_analysis.py | 39 +++++++++++++++++++++++++++++++++++++++
 tests/cli/get/test_cli_get_family.py   | 37 +++++++++++++++++++++++++++++++++++++
 6 files changed, 113 insertions(+), 6 deletions(-)
 create mode 100644 tests/cli/get/test_cli_get_analysis.py
 create mode 100644 tests/cli/get/test_cli_get_family.py
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
Switched to branch 'gisaid'
Your branch is behind 'origin/gisaid' by 47 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
/home/maya.brandi
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/maya.brandi
done.
Log deploy... done.
Running --version on each Clinical-Genomics developed tool to verify that all requirements are met:
pytest-cov 2.10.1 has requirement pytest>=4.6, but you have pytest 4.2.1.
multiqc 1.8 has requirement matplotlib<3.1.2,>=2.1.1, but you have matplotlib 3.3.3.
demux 5.4.5 has requirement importlib-metadata<4.0.0,>=3.1.1, but you have importlib-metadata 2.1.1.
balsamic 7.1.1 has requirement click>=7.1.2, but you have click 6.7.
/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
cg, version 20.22.0
housekeeper, version 3.2.0
demux, version 5.6.1
cglims, version 1.3.4
[0|0|521] 2d [maya.brandi@hasta:~] [P_main] 2m29s $ bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-servers-stage.sh gisaid
on the host hasta.scilifelab.se, continue
repository is clean
remote: Enumerating objects: 11, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 11 (delta 2), reused 4 (delta 1), pack-reused 0
Unpacking objects: 100% (11/11), done.
From github.com:Clinical-Genomics/servers
   721062c..02a70cf  gisaid     -> origin/gisaid
   08b720b..96d0a91  master     -> origin/master
 * [new tag]         v5.33.0    -> v5.33.0
Switched to branch 'gisaid'
Your branch is behind 'origin/gisaid' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)
Updating 721062c..02a70cf
Fast-forward
 config/hasta.scilifelab.se/cg.yaml | 4 ++++
 1 file changed, 4 insertions(+)
/home/maya.brandi
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/maya.brandi
done.
Log deploy... done.
[0|0|521] 2d [maya.brandi@hasta:~] [P_main] 6s $ us
[0|0|521] 2d [maya.brandi@hasta:~] [S_main] $ python /home/proj/stage/servers/resources/gisaid_uploader.py CoV authenticate --cid TEST-EA76875B00C3
Enter Username:  i.sylvin
Enter Password:  
valid until: Wed Jul 28 16:42:40 2021
user authenticated
[0|0|521] 2d [maya.brandi@hasta:~] [S_main] 14s $ cg -c /home/proj/stage/servers/config/hasta.scilifelab.se/cg-stage.yaml upload gisaid frankhusky
/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
----------------- UPLOAD ----------------------
----------------- GISAID -------------------
/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Demux.samples to refer to viewonly relationship Sample.demuxes should include sync_backref=False set on the Demux.samples relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Sample.demuxes to refer to viewonly relationship Demux.samples should include sync_backref=False set on the Sample.demuxes relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A1]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A2]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A3]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A4]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A5]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A6]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A7]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A8]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A9]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A10]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A11]
Fetching files from version 35889
Fetch latest version from bundle frankhusky
Fetching files with tags in [consensus,ACC7703A12]
Fetching files from version 35889
Running command /home/proj/stage/servers/resources/gisaid_uploader.py CoV upload --csv /home/maya.brandi/frankhusky.csv --fasta /home/maya.brandi/frankhusky.fasta
gisaid stdout:
hCoV-19/Sweden/01_SE100_21CS500657TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500658TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500659TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500660TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500661TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500662TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500663TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500664TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500665TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500666TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500667TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
hCoV-19/Sweden/01_SE100_21CS500668TEST/2020-11-22; validation_error; {"covv_virus_name": "already exists"}
submissions uploaded: 0
submissions failed: 12

````

👍
